keras_contrib/callbacks/dead_relu_detector.py

```diff
@@ -44,17 +48,43 @@ def get_relu_activations(self):
         layer_outputs = [func(list_inputs)[0] for func in funcs]
         for layer_index, layer_activations in enumerate(layer_outputs):
             if self.is_relu_layer(self.model.layers[layer_index]):
-                yield [layer_index, layer_activations]
+                layer_name = self.model.layers[layer_index].name
+                # layer_weight is a list [W] (+ [b])
+                layer_weight = self.model.layers[layer_index].get_weights()
+                # with kernel and bias, the weights are saved as a list [W, b]. If only weights, it is [W]
+                assert type(layer_weight) == list
+                layer_weight_shape = np.shape(layer_weight[0])
+                yield [layer_index, layer_activations, layer_name, layer_weight_shape]
 
     def on_epoch_end(self, epoch, logs={}):
         for relu_activation in self.get_relu_activations():
-            layer_index, activation_values = relu_activation
-            total_neurons = activation_values.shape[-1]
-            dead_neurons = np.sum(activation_values == 0)
-            dead_neurons_share = dead_neurons / total_neurons
+            layer_index, activation_values, layer_name, layer_weight_shape = relu_activation
+
+            shape_act = activation_values.shape
+
+            weight_len = len(layer_weight_shape)
+            act_len = len(shape_act)
+
+            # should work for both Conv and Flat
+            if K.backend() == 'tensorflow':
+                # features in last axis
+                axis_filter = -1
+            elif K.backend() == 'theano':
+                # features before the convolution axis, for weight_len the input and output have to be subtracted
+                axis_filter = -1 - (weight_len - 2)
+            else:
+                raise ValueError('Unknown backend: {}'.format(K.backend()))
```

Comment:
@infelane 
At the very least there is also NLTK backend (Microsoft's Natual Language Toolkit)
There are also some third-party back-ends, like plaidml: https://github.com/plaidml/plaidml
There may be more back-ends coming in the future

The bottom line is that ideally you should not care how the back-end is implemented in `keras`
Maybe there is a way to ask backend about axis order?

Reply:
K.image_data_format() should do the trick
Done

Reply:
I think this was a necessary change. Even between TF and Theano you can still adjust the channel order. Making the previous check prone to fail if you were going to adjust the channel order manually.